tests/test_router_direct_custom_quote_in_given_out.py

```diff
@@ -0,0 +1,87 @@
+import time
+import pytest
+
+import conftest
+from sqs_service import *
+from quote import *
+from quote_response import *
+from rand_util import *
+from e2e_math import *
+from decimal import *
+from constants import *
+from util import *
+from route import *
+
+# Arbitrary choice based on performance at the time of test writing
+EXPECTED_LATENCY_UPPER_BOUND_MS = 15000
+
+# Test suite for the /router/custom-direct-quote endpoint
+# Test runs tests for exact amount out quotes.
+class TestExactAmountOutDirectCustomQuote:
+    @pytest.mark.parametrize("pair", conftest.create_coins_from_pairs(conftest.create_no_dupl_token_pairs(conftest.choose_tokens_liq_range(num_tokens=10, min_liq=500_000, exponent_filter=USDC_PRECISION)), USDC_PRECISION, USDC_PRECISION + 3), ids=id_from_swap_pair)
+    def test_get_custom_direct_quote(self, environment_url, pair):
+        denom_in = pair['out_denom']
+        coin = Coin(pair['token_in']['denom'], pair['token_in']['amount_str'])
+        token_out  = pair['token_in']['amount_str'] + pair['token_in']['denom']
+
+        # Get the optimal quote for the given token pair
+        # Direct custom quote does not support multiple routes, so we request single/multi hop pool routes only
+        optimal_quote = ExactAmountOutQuote.run_quote_test(environment_url, token_out, denom_in, False, True, EXPECTED_LATENCY_UPPER_BOUND_MS)
+
+        pool_id = ','.join(map(str, optimal_quote.get_pool_ids()))
+        denoms_in = ','.join(map(str, optimal_quote.get_token_in_denoms()))
+
+        quote = self.run_quote_test(environment_url, token_out, denoms_in, pool_id, EXPECTED_LATENCY_UPPER_BOUND_MS)
+
+        # All tokens have the same default exponent, resulting in scaling factor of 1.
+        spot_price_scaling_factor = 1
+
+        # Compute expected base out quote spot price
+        # First, get the USD price of each denom, and then divide to get the expected spot price
+        in_base_usd_quote_price = conftest.get_usd_price_scaled(denom_in)
+        out_base_usd_quote_price = conftest.get_usd_price_scaled(coin.denom)
+        expected_in_base_out_quote_price = out_base_usd_quote_price / in_base_usd_quote_price 
+
+        # Compute expected token out
+        expected_token_in = int(coin.amount) * expected_in_base_out_quote_price
```

Comment:
Would it be possible to create a helper to abstract this to reduce code duplication?

I think we could reuse that helper [here](https://vscode.dev/github/osmosis-labs/sqs/blob/SQS-303-e2e/tests/test_router_quote_in_given_out.py#L91-L98)

Similar abstraction could be added for "out given in"

Reply:
Moved such logic to `calculate_expected_base_out_quote_spot_price` methods for given out and given in respectively. :rocket: